src/firstboot/firstboot.c

```diff
@@ -582,23 +586,57 @@ static int prompt_root_password(void) {
         return 0;
 }
 
-static int write_root_shadow(const char *path, const struct spwd *p) {
-        _cleanup_fclose_ FILE *f = NULL;
+static int write_root_shadow(const char *shadow_path, const struct spwd *p) {
+        _cleanup_fclose_ FILE *original = NULL, *shadow = NULL;
+        _cleanup_(unlink_and_freep) char *shadow_tmp = NULL;
+        struct spwd *sp = NULL;
         int r;
 
-        assert(path);
-        assert(p);
+        r = fopen_temporary_label("/etc/shadow", shadow_path, &shadow, &shadow_tmp);
+        if (r < 0)
+                return r;
+
+        original = fopen(shadow_path, "re");
+        if (original) {
+
+                r = sync_rights(original, shadow_tmp);
+                if (r < 0)
+                        return r;
+
+                while ((r = fgetspent_sane(original, &sp)) > 0) {
+
+                        if (streq(sp->sp_namp, p->sp_namp))
+                                r = putspent_sane(p, shadow); // Replace existing matching entry.
```

Comment:
as per coding style we use `//` comments for "FIXME" style stuff, and `/* */` style comments for explanatory stuff. This is the latter hence should use `/* */`